### PR TITLE
chore(config): Convert top-level transforms enum to typetag

### PR DESCRIPTION
**Pull request by bruceg**
This replaces the top-level `enum Transforms` into a boxed transform type. Serialization and deserialization are handled by `typetag`, and the `configurable_component` macro is enhanced to build out the necessary table entries to generate the schema bits dynamically from all of the components that are compiled into the current configuration.

Note that this same approach is now possible for both the sources and sinks as well. This focuses on the transforms as the smallest step to prove it out for the other component types. Making it work for either sources or sinks should be able to be scoped down to just the actual sources and sinks and not need to touch support code like this does.

### CODE REVIEW

File: src/config/transform.rs [hunk 1]
```diff
@@ -142,7 +171,8 @@ impl TransformContext {
 /// Generalized interface for describing and building transform components.
 #[async_trait]
 #[enum_dispatch]
```
Comment (tobz) [inline src/config/transform.rs:173]: Should be able to drop this line now.

**Review by tobz — APPROVED** [review]
Generally looks good to me, with a few nits/questions.

File: lib/vector-config/src/component/description.rs [hunk 2]
```diff
@@ -67,4 +84,47 @@ where
...
+            .collect::<Result<_, _>>()?;
+        Ok(schema::generate_one_of_schema(&subschemas))
+    }
+
+    /// Generate a schema object for this description.
+    fn generate_schema(
+        &self,
+        gen: &RefCell<SchemaGenerator>,
+    ) -> Result<SchemaObject, GenerateError> {
+        let tag_schema = schema::generate_internal_tagged_variant_schema("type".to_string(), {
```
[the remark this answers is not in the document]
Comment (bruceg) [reply lib/vector-config/src/component/description.rs:104]: I'm open to that, but this seemed the best first path to solving it.

File: lib/vector-config/Cargo.toml [hunk 3]
```diff
@@ -23,6 +23,7 @@ serde_json = { version = "1.0", default-features = false, features = ["std"] }
 serde_with = { version = "2.2.0", default-features = false, features = ["std"] }
 snafu = { version = "0.7.4", default-features = false }
 toml = { version = "0.7.2", default-features = false }
+typetag = { version = "0.2.5", default-features = false }
```
[the remark this answers is not in the document]
Comment (bruceg) [reply lib/vector-config/Cargo.toml]: I think so, yes, I will have put it in thinking it was needed to collect… something, but it isn't needed.

**Pull request merged**